File: app/src/main/java/com/adisa/diningplus/db/entities/Hall.java
package com.adisa.diningplus.db.entities;

import androidx.annotation.NonNull;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.Serializable;

@Entity(tableName = "halls")
public class Hall implements Serializable {
    @PrimaryKey
    @NonNull
    public String id;
    public String name;
    public String nickname;
    public boolean open;
    public int occupancy;
    public double latitude;
    public double longitude;
    public String address;
    public String phone;

    public static Hall fromJSON(JSONObject raw) throws JSONException {
        Hall hall = new Hall();
        hall.id = raw.getString("id");
        hall.name = raw.getString("name");
        hall.nickname = raw.getString("nickname");
        hall.open = raw.optBoolean("open");
        hall.occupancy = raw.optInt("occupancy");
        hall.latitude = raw.optDouble("latitude");
        hall.longitude = raw.optDouble("longitude");
        hall.address = raw.optString("address");
        hall.phone = raw.optString("phone");
        return hall;
    }
}

File: app/src/main/java/com/adisa/diningplus/db/entities/Hall.kt
package com.adisa.diningplus.db.entities

import android.location.Location
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import org.json.JSONObject

@Entity(tableName = "halls")
data class Hall(val raw: JSONObject) {
    @PrimaryKey
    var id: String = raw.getString("id")
    var name: String = raw.getString("name")
    var nickname: String = raw.getString("nickname")
    var open: Boolean = raw.optBoolean("open")
    var occupancy: Int = raw.optInt("occupancy")
    var latitude: Double = raw.optDouble("latitude")
    var longitude: Double = raw.optDouble("longitude")
    var address: String = raw.optString("address")
    var phone: String = raw.optString("phone")

    @Ignore
    var distance = 0.0
    fun setDistance(location: Location?) {
        if (location != null) {
            val loc = Location("")
            loc.latitude = latitude
            loc.longitude = longitude
            // Convert m -> km
            distance = location.distanceTo(loc) / 1000.toDouble()
        }
    }
}